### PR TITLE
Run benchmark trials in random order #11805

### PR DESCRIPTION
**Pull request by mannanali413**
This Pull Request Address Issue #11805 .  

The Benchmarks are randomly picked up and run. This ensures that no single benchmark is impacted in case other processes grab the CPU.

### CODE REVIEW

File: scripts/bench/runner.js [hunk 1]
```diff
@@ -30,29 +33,66 @@ const benchmarkFilter = argv.benchmark;
...
+  let filteredBenchmarkNames = benchmarkNames;
+  if (benchmarkFilter) {
+    filteredBenchmarkNames = benchmarkNames.filter(
+      name => name.indexOf(benchmarkFilter) !== -1
+    );
+  }
+  let benchmarkProgressObj = initializeBenchmarkProgressObj(
+    filteredBenchmarkNames
+  );
+  let getRandomBenchmarkName = uniqueRandomArray(filteredBenchmarkNames);
```
Comment (gaearon) [inline scripts/bench/runner.js]: Can we use something well-known like Lodash for this?

File: scripts/bench/runner.js [hunk 2]
```diff
@@ -30,29 +33,66 @@ const benchmarkFilter = argv.benchmark;
 const headless = argv.headless;
 const skipBuild = argv['skip-build'];
 
+function terminationReached(benchmarkProgressObj) {
```
Comment (gaearon) [inline scripts/bench/runner.js]: Maybe `shouldTerminate`?

File: scripts/bench/runner.js [hunk 3]
```diff
@@ -30,29 +33,66 @@ const benchmarkFilter = argv.benchmark;
...
+  let terminated = populatedNames.reduce((base, benchmarkName) => {
+    base =
+      base && benchmarkProgressObj[benchmarkName] === BENCHMARK_TRIAL_RUN_LIMIT;
+    return base;
+  }, true);
+  return terminated;
+}
+
+function initializeBenchmarkProgressObj(benchmarkNames) {
+  let progressObj = {};
```
Comment (gaearon) [inline scripts/bench/runner.js]: Nit: Maybe `progressState`? Not a fan of `obj` suffix.

File: scripts/bench/runner.js [hunk 4]
```diff
@@ -30,29 +33,66 @@ const benchmarkFilter = argv.benchmark;
...
         chalk.gray(`- Building benchmark "${chalk.white(benchmarkName)}"...`)
       );
       await buildBenchmark(reactPath, benchmarkName);
       console.log(
         chalk.gray(`- Running benchmark "${chalk.white(benchmarkName)}"...`)
       );
-      results[benchmarkName] = await runBenchmark(benchmarkName, headless);
+      let benchmarkResult = await runBenchmark(benchmarkName, headless);
+      results[benchmarkName].runs.push(benchmarkResult);
+      benchmarkProgressObj[benchmarkName] += 1;
```
Comment (gaearon) [inline scripts/bench/runner.js]: Just `++`

**Comment by gaearon** [comment]
Nice, thanks for looking at this! Left a few nits.

Can you also summarize changes to control flow in a few paragraphs in a comment? I don't remember how the original worked so it would be helpful for me to feel confident about changes.

**Comment by mannanali413** [comment]
@gaearon thanks for reviewing this . 

To get more clarity on the interleaving I need some help from your side. Suppose that we have to run trials for **n** tasks(_say **T<sub>1</sub>, T<sub>2</sub>,... T<sub>n</sub>**_) . 
The local benchmark for **i<sup>th</sup>** task **T<sub>i</sub>** is denoted as **LT<sub>i</sub>** and the remote benchmark for the same task is denoted as **RT<sub>i</sub>** . 

For each task **T<sub>i</sub>** benchmarks are run **MAX_NUMBER_OF_ITERATIONS** for local and remote benchmarks. Hence, it could also be said that the benchmark tests are run iteratively till the termination count is reached. Let's suppose that at each iteration(_j_) a random number **r<sub>j</sub>** is generated such that **1 <= r<sub>j</sub> <= n**. The random number generated at iteration _j_ is different from that generated at iteration _j+1_, thus **r<sub>j</sub> is not equal to r<sub>j+1</sub>**

Using the random number generated at each iteration and interleaving the remote and the local tasks would yield sequences of the following order . 

**Sequence Type 1**
- _LT<sub>r<sub>j</sub></sub>_ (Local Benchmark  For Task<sub>r<sub>j</sub></sub>)
- _RT<sub>r<sub>j</sub></sub>_ (Remote Task for the same Task<sub>r<sub>j</sub></sub>)
- _LT<sub>r<sub>j+1</sub></sub>_ (Local Benchmark  For Task<sub>r<sub>j+1</sub></sub>)
- _RT<sub>r<sub>j+1</sub></sub>_ (Remote Task for the same Task<sub>r<sub>j+1</sub></sub>)
- ....so on

**Sequence Type 2**
- _LT<sub>r<sub>j</sub></sub>_ (Local Benchmark  For Task<sub>r<sub>j</sub></sub>)
- _RT<sub>r<sub>j+1</sub></sub>_ (Remote Task for the same Task<sub>r<sub>j+1</sub></sub>)
- _LT<sub>r<sub>j+2</sub></sub>_ (Local Benchmark  For Task<sub>r<sub>j+2</sub></sub>)
- _RT<sub>r<sub>j+3</sub></sub>_ (Remote Task for the same Task<sub>r<sub>j+3</sub></sub>)
- ....so on

In _Sequence Type 1_, the local and remote benchmarks for a Task are run one after the other. On the other hand, _Sequence Type 2_ ensures that no two benchmarks(be they local or remote) are run for the same task simultaneously. Of the two sequences, @gaearon which one do u suppose would produce a non-skewed result? 

In my opinion generating _Sequence Type 2_ for producing benchmark reports would be a better approach.

**Comment by gaearon** [comment]
Hmm, I’m not sure. I trust you to figure this out, try different approaches, and tell me what works better 🙂

**Comment by cpojer** [comment]
Hey! It seems like this PR got stale and is now really outdated. If this is something that still makes sense to the latest master version of React, please send a new PR with this change :)

**Pull request closed without merging**